Add `ToCBOR`/`FromCBOR` instances for `Genesis` config types

Soupstraw opened the pull request:
# Description

Needed `FromCBOR` and `ToCBOR` instances for:

* FromByronTranslationContext
* NoGenesis {AllegraEra, MaryEra, BabbageEra}
* ConwayGenesis
* Network
* ActiveSlotCoeff

Closes https://github.com/IntersectMBO/cardano-ledger/issues/4893

<!-- Add your description here, if it fixes a particular issue please provide a
[link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=)
to the issue. -->

# Checklist

- [ ] Commits in meaningful sequence and with useful messages
- [ ] Tests added or updated when needed
- [ ] `CHANGELOG.md` files updated for packages with externally visible changes<br>
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] Versions updated in `.cabal` and `CHANGELOG.md` files when necessary, according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] Version bounds in `.cabal` files updated when necessary<br>
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code formatted (use `scripts/fourmolize.sh`)
- [x] Cabal files formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

Soupstraw committed:
WIP

jasagredo commented:
In case this PR is for https://github.com/IntersectMBO/cardano-ledger/issues/4893, in consensus we actually need `ToCBOR` and `FromCBOR` instances, which I think enable deriving `Enc/DecCBOR` but not the other way round.

If the PR is not for that issue feel free to ignore my comment

teodanciu committed:
Add a few `ToCBOR` and `FromCBOR` instances

and modify `EncCBOR` and `DecCBOR` to use them when appropriate.
`ConwayGenesis`, `NoGenesis`, `Network`, `ActiveSlotCoeff`,
`FromByronTranslationContext`

teodanciu committed:
Define some decoding/encoding functions to `Binary.Plain`

And reuse them in `Decoder` when possible: `decodeRationalWithTag`,
`encodeRatioWithTag`,`assertTag`, `decodeTagMaybe`

teodanciu commented on libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs:
```diff
@@ -344,6 +346,20 @@ fromRatioBoundedRatio ratio
+  (FromCBOR a, Bounded (BoundedRatio b a), Bounded a, Integral a, Typeable b, Show a) =>
+  FromCBOR (BoundedRatio b a)
+  where
+  fromCBOR = do
```
This should be equivalent to `EncCBOR (BoundedRatio)`, in so far as  `decodeCollectionWithLen` [from cardano-binary](https://github.com/IntersectMBO/cardano-base/blob/56ed3e31560f5c9c34c2f602ccb278493a0d49ae/cardano-binary/src/Cardano/Binary/FromCBOR.hs#L584) is equivalent to `decodeCollectionWithLen` from [Decoder](https://github.com/IntersectMBO/cardano-ledger/blob/edb2ee48409871eae0687692994c11d38ffd261a/libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Decoder.hs#L736).
I think they are, but I hesitate to derive `DecCBOR` from this implementation.

lehins replied:
Feel free to derive it, since I can confirm that those implementations are equivalent. It would be more confusing to two have identical alternative implementations

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Plain.hs:
```diff
@@ -142,3 +148,53 @@ decodeEnumBounded = do
+      "Expecteg tag " <> show tagExpected <> " but got tag " <> show tagReceived
+{-# INLINE assertTag #-}
+
+decodeTagMaybe :: Decoder s (Maybe Word64)
```
Can you please reuse this implementation for versioned decoder in order to reduce duplication: https://github.com/IntersectMBO/cardano-ledger/blob/eec723f451f6b100b281570e2a672435d07c1b9b/libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Decoder.hs#L1206

teodanciu replied:
Sure, should have thought of it! Thanks

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Plain.hs:
```diff
@@ -142,3 +148,53 @@ decodeEnumBounded = do
+    <> encodeNumeric (denominator r)
+
+assertTag :: Word -> Decoder s ()
+assertTag tagExpected = do
```
Ditto. Reuse this implementation in https://github.com/IntersectMBO/cardano-ledger/blob/eec723f451f6b100b281570e2a672435d07c1b9b/libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Decoding/Decoder.hs#L1224

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/Genesis.hs:
```diff
@@ -16,3 +23,13 @@ class Era era => EraGenesis era where
+  toCBOR _ = toCBOR ()
+
+instance Era era => FromCBOR (NoGenesis era) where
+  fromCBOR = fromCBOR @() *> pure NoGenesis
```
Slight simplification
```suggestion
  fromCBOR = NoGenesis <$ fromCBOR @()
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs:
```diff
@@ -632,16 +650,17 @@ instance NoThunks ActiveSlotCoeff
+instance ToCBOR ActiveSlotCoeff where
+  toCBOR x@(ActiveSlotCoeff _ _) =
+    let ActiveSlotCoeff {..} = x
+     in toCBOR unActiveSlotVal
```
With previous implementation it was obvious that a field was ignored. Current implementation makes it totally not obvious, therefore a comment is a must
```suggestion
     -- `unActiveSlotLog` is not encoded, since it can be derived from `unActiveSlotVal`
     in toCBOR unActiveSlotVal
```

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Plain.hs:
```diff
@@ -142,3 +148,53 @@ decodeEnumBounded = do
+      when (d == 0) (fail "Denominator cannot be zero")
+      pure $! n % d
+    _ -> cborError $ DecoderErrorSizeMismatch "Rational" 2 numValues
+{-# INLINE decodeRationalWithoutTag #-}
```
Plain Rational decoder without a tag will never be needed

```suggestion
decodeRationalWithTag :: Decoder s Rational
decodeRationalWithTag = do
  assertTag 30
  (numValues, values) <- decodeCollectionWithLen decodeListLenOrIndef decodeInteger
  case values of
    [n, d] -> do
      when (d == 0) (fail "Denominator cannot be zero")
      pure $! n % d
    _ -> cborError $ DecoderErrorSizeMismatch "Rational" 2 numValues
{-# INLINE decodeRationalWithTag #-}
```

lehins commented on libs/cardano-ledger-binary/src/Cardano/Ledger/Binary/Plain.hs:
```diff
@@ -142,3 +148,53 @@ decodeEnumBounded = do
+encodeRatioNoTag encodeNumeric r =
+  encodeListLen 2
+    <> encodeNumeric (numerator r)
+    <> encodeNumeric (denominator r)
```
Ditto
```suggestion
encodeRatioWithTag encodeNumeric r =
  encodeTag 30
    <> encodeListLen 2
    <> encodeNumeric (numerator r)
    <> encodeNumeric (denominator r)
```

lehins commented on libs/cardano-ledger-core/src/Cardano/Ledger/BaseTypes.hs:
```diff
@@ -792,15 +811,22 @@ word8ToNetwork e
+              Nothing -> cborError $ DecoderErrorCustom "Network" "Unknown network id"
+              Just n -> pure n
+          )
+        . word8ToNetwork
```
Sorry, but it looks too ugly. This will be more concise and more readable:
```suggestion
  fromCBOR = do
    w8 <- decodeWord8
    case word8ToNetwork w8 of
      Nothing -> cborError $ DecoderErrorCustom "Network" "Unknown network id"
      Just n -> pure n
```

lehins reviewed (approved):
Looks good.
Just a few minor suggestions.
